src/main/java/eu/dissco/core/digitalmediaobjectprocessor/service/ProcessingService.java

```diff
@@ -66,13 +70,29 @@ public List<DigitalMediaObjectRecord> handleMessage(List<DigitalMediaObjectTrans
     return results;
   }
 
-  private Set<DigitalMediaObjectTransferEvent> removeDuplicatesInBatch(
-      List<DigitalMediaObjectTransferEvent> events) {
-    var uniqueSet = new HashSet<DigitalMediaObjectTransferEvent>();
+  private void checkIfDigitalSpecimenIdExists(List<DigitalMediaObjectEvent> events)
+      throws DigitalSpecimenNotFoundException {
+    var digitalSpecimenIds = events.stream().map(DigitalMediaObjectEvent::digitalMediaObject)
+        .map(DigitalMediaObject::digitalSpecimenId).toList();
+    var existingDigitalSpecimenIds = digitalSpecimenRepository.getExistingSpecimen(
+        digitalSpecimenIds);
+    if (!digitalSpecimenIds.equals(existingDigitalSpecimenIds)) {
+      var nonExistingIds = new ArrayList<>(digitalSpecimenIds);
+      nonExistingIds.removeAll(existingDigitalSpecimenIds);
+      log.error("Digital specimen ids {} do not exist in database", nonExistingIds);
+      throw new DigitalSpecimenNotFoundException(
+          "Digital specimen ids: " + nonExistingIds + " do not exist in database");
+    }
+  }
+
+  private Set<DigitalMediaObjectEvent> removeDuplicatesInBatch(
+      List<DigitalMediaObjectEvent> events) {
+    var uniqueSet = new LinkedHashSet<DigitalMediaObjectEvent>();
     var map = events.stream()
         .collect(
-            Collectors.groupingBy(event -> getMediaUrl(event.digitalMediaObject().attributes())));
-    for (Entry<String, List<DigitalMediaObjectTransferEvent>> entry : map.entrySet()) {
+            Collectors.groupingBy(
+                event -> event.digitalMediaObject().attributes().getAcAccessUri()));
+    for (Entry<String, List<DigitalMediaObjectEvent>> entry : map.entrySet()) {
       if (entry.getValue().size() > 1) {
         log.warn("Found {} duplicates in batch for id {}", entry.getValue().size(), entry.getKey());
         for (int i = 0; i < entry.getValue().size(); i++) {
```

Comment:
Because we grap batches from the queue. It could be that we ingested an item (place 1) and then some other things and then ingested the same item but now updated (place 300). This would create conflicts as we now have the same item twice, one original and one updated. So before we continue we first remove the updated one from this batch and push it back to the queue. We can then safely process the first item and then pickup the updated one in a later batch. Still things could go wrong but this should help a bit

Reply:
Thanks for explaining!